src/main/java/Grammar.java

```diff
@@ -0,0 +1,143 @@
+            int counter = 0;
+            for (String line : lines) {
+                if (counter == 0) {
+                    String[] tokens = line.split(", ");
+                    nonterminals.addAll(List.of(tokens).stream().map(token -> token.toUpperCase(Locale.ROOT)).collect(Collectors.toList()));
+
+                } else if (counter == 1) {
+                    String[] tokens = line.split(", ");
+                    terminals.addAll(List.of(tokens));
+
+                } else if (counter == 2) {
+                    starterNonterminal = line.strip().toUpperCase(Locale.ROOT);
+                } else {
+                    String[] tokens = line.split("::=");
+                    String lhs = tokens[0];
+                    String rhs = tokens[1];
+                    String[] lhsTerminals = lhs.split("[<> ]");
+                    List<String> key = new ArrayList<>();
+
+                    for (String terminal : lhsTerminals) {
+                        if (!terminal.isEmpty()) {
+                            key.add(terminal.toUpperCase(Locale.ROOT));
+                        }
+                    }
+
+                    String[] rhsProductions = rhs.split("\\|");
+                    for(String rhss : rhsProductions) {
+                        if (!productionsSet.productions.containsKey(key)) {
+                            productionsSet.productions.put(key, new ArrayList<>());
+                        }
+                        List<String> value = new ArrayList<>();
+                        String[] rhsTokens = rhss.split(" ");
+                        for (String token : rhsTokens) {
+                            if (!token.isEmpty()) {
+                                if (token.charAt(0) == '<' && token.charAt(token.length() - 1) == '>') {
+                                    token = token.substring(1, token.length() - 1);
+                                    value.add(token.toUpperCase(Locale.ROOT));
+                                } else {
+                                    token = token.strip();
+                                    value.add(token);
+                                }
+                            }
+                        }
+
+                        List<List<String>> list = productionsSet.productions.get(key);
+                        list.add(value);
+                    }
+                }
+                counter++;
+            }
+        } catch (IOException ex) {
+            System.out.format("I/O error: %s%n", ex);
+        }
+    }
+
+    void printNonTerminals() {
+        System.out.println(nonterminals);
+    }
+
+    void printTerminals() {
+        System.out.println(terminals);
+    }
+
+    void printProductions() {
+        System.out.println(productionsSet);
+    }
+
+    List<Production> printProductionsForGivenNonTerminal (String nonterminal) {
+        List<Production> productions = new ArrayList<>();
+        if (!this.nonterminals.contains(nonterminal)) {
```

Comment:
on line 27, when adding a nonterminal to the set of nonterminals, it is added as uppercase, whilst here it is searched by its original name (which is not necessarily an uppercase string, i.e. "program" from g2). Maybe try and search for the uppercase of the nonterminal given as parameter?